crates/types/src/storage.rs

```diff
@@ -395,6 +399,126 @@ pub fn decode_from_flexbuffers<T: DeserializeOwned, B: Buf>(
     }
 }
 
+/// [`ArcVec`] mainly used by `message` types to improve
+/// cloning of messages.
+///
+/// It can replace [`Vec<T>`] most of the time in all structures
+/// that need to be serialized over the wire.
+///
+/// Internally it keeps the data inside an [`Arc<[T]>`]
+#[derive(Debug)]
+pub struct ArcVec<T> {
+    inner: Arc<[T]>,
+}
+
+impl<T> Deref for ArcVec<T> {
+    type Target = [T];
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<T> serde::Serialize for ArcVec<T>
+where
+    T: serde::Serialize,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut seq = serializer.serialize_seq(Some(self.len()))?;
+        for elem in self.iter() {
+            seq.serialize_element(elem)?;
+        }
+
+        seq.end()
+    }
+}
+
+impl<'de, T> serde::Deserialize<'de> for ArcVec<T>
+where
+    T: serde::Deserialize<'de>,
+{
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_seq(ArcVecVisitor::default())
+    }
+}
+
+struct ArcVecVisitor<T> {
+    _phantom: PhantomData<T>,
+}
+
+impl<T> Default for ArcVecVisitor<T> {
+    fn default() -> Self {
+        Self {
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<'de, T> serde::de::Visitor<'de> for ArcVecVisitor<T>
+where
+    T: serde::Deserialize<'de>,
+{
+    type Value = ArcVec<T>;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        write!(formatter, "expecting an array")
+    }
+
+    fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+    where
+        A: serde::de::SeqAccess<'de>,
+    {
+        let mut vec: Vec<T> = Vec::with_capacity(seq.size_hint().unwrap_or_default());
+        while let Some(value) = seq.next_element()? {
+            vec.push(value);
+        }
+
+        Ok(vec.into())
```

Comment:
Once https://github.com/rust-lang/rust/pull/129401 lands with Rust 1.82, we can avoid the extra allocation of a `Vec` here and directly produce into the Arc<[T]> slice.